File: day-2/Sorting-algos/sort-an-array/mergesort.js
function sortArray(arr) {
    if(arr.length <= 1) {
      return arr;
    }
    let mid = Math.floor(arr.length / 2)
    let left = arr.slice(0, mid)
    let right = arr.slice(mid)

    return merge(sortArray(left), sortArray(right));
  }

  function merge(left, right) {
      let merged = []
      let i = 0
      let j = 0
      while(i < left.length && j < right.length){
          if(left[i] < right[j]) {
              merged.push(left[i])
              i++
          } else {
              merged.push(right[j])
              j++
          }
      }
      while(i < left.length) {
          merged.push(left[i])
          i++
      }
      while(j < right.length) {
          merged.push(right[j])
          j++
      }

      return merged
  }

  function sortArray(arr) {
    if(arr.length <= 1){
        return arr
    }

    let pivot = Math.floor(arr.length/2)
    let left = arr.slice(0, pivot)
    let right = arr.slice(pivot)
    return merge(sortArray(left), sortArray(right))
}

function merge(left, right) {
    let merged = [];
    let i = 0;
    let j = 0;
    while(i < left.length && j < right.length){
        if(left[i] < right[j]){
            merged.push(left[i])
            i++
        } else {
            merged.push(right[j])
            j++
        }
    }

    while(i < left.length){
        merged.push(left[i])
        i++
    }

    while(j < right.length){
        merged.push(right[j])
        j++
    }

    return merged
}

File: day-2/Sorting-algos/sort-an-array/mergesort.py
def sortArray(self, nums: List[int]) -> List[int]:
        if len(nums) < 2: return nums

        mid = len(nums) // 2
        left = nums[0:mid]
        right = nums[mid::]
        return merge(self.sortArray(left), self.sortArray(right))

def merge(left, right):
        merged = []
        i = 0
        j = 0
        while i < len(left) and j < len(right):
            if left[i] < right[j]:
                merged.append(left[i])
                i = i + 1
            else:
                merged.append(right[j])
                j = j + 1
        while i < len(left):
            merged.append(left[i])
            i = i + 1
        while j < len(right):
            merged.append(right[j])
            j = j + 1

        return merged
